File: client/src/projectCreationPathValidation.ts
export const projectCreationPathValidationReasons = {
  projectNameEmpty: 'projectNameEmpty',
  projectNameIllFormedUnicode: 'projectNameIllFormedUnicode',
  projectNameDotSegment: 'projectNameDotSegment',
  projectNameContainsPathSeparator: 'projectNameContainsPathSeparator',
  projectNameContainsWindowsInvalidCharacter: 'projectNameContainsWindowsInvalidCharacter',
  projectNameContainsUnicodeControlCharacter: 'projectNameContainsUnicodeControlCharacter',
  projectNameHasLeadingOrTrailingWhitespace: 'projectNameHasLeadingOrTrailingWhitespace',
  projectNameEndsWithDot: 'projectNameEndsWithDot',
  projectNameUsesReservedDeviceName: 'projectNameUsesReservedDeviceName',
  excelPathContainsUnsupportedCharacter: 'excelPathContainsUnsupportedCharacter',
  excelPathTooLong: 'excelPathTooLong'
} as const;

export type ProjectCreationPathValidationReason =
  typeof projectCreationPathValidationReasons[keyof typeof projectCreationPathValidationReasons];

export interface ProjectCreationPathValidationResult {
  readonly isValid: boolean;
  readonly reason: ProjectCreationPathValidationReason | null;
}

export const maximumExcelWorkbookPathUtf16CodeUnitLength = 218;

const reservedDeviceNames = new Set([
  'CON',
  'PRN',
  'AUX',
  'NUL',
  'COM1',
  'COM2',
  'COM3',
  'COM4',
  'COM5',
  'COM6',
  'COM7',
  'COM8',
  'COM9',
  'COM¹',
  'COM²',
  'COM³',
  'LPT1',
  'LPT2',
  'LPT3',
  'LPT4',
  'LPT5',
  'LPT6',
  'LPT7',
  'LPT8',
  'LPT9',
  'LPT¹',
  'LPT²',
  'LPT³'
]);

export function validateProjectName(candidate: string): ProjectCreationPathValidationResult {
  if (candidate.length === 0) {
    return rejected(projectCreationPathValidationReasons.projectNameEmpty);
  }

  if (!isWellFormedUtf16(candidate)) {
    return rejected(projectCreationPathValidationReasons.projectNameIllFormedUnicode);
  }

  if (candidate === '.' || candidate === '..') {
    return rejected(projectCreationPathValidationReasons.projectNameDotSegment);
  }

  if (candidate.includes('/') || candidate.includes('\\')) {
    return rejected(projectCreationPathValidationReasons.projectNameContainsPathSeparator);
  }

  if (/[<>:"|?*]/u.test(candidate)) {
    return rejected(projectCreationPathValidationReasons.projectNameContainsWindowsInvalidCharacter);
  }

  if (containsUnicodeControlCodeUnit(candidate)) {
    return rejected(projectCreationPathValidationReasons.projectNameContainsUnicodeControlCharacter);
  }

  if (isContractWhitespace(candidate.charCodeAt(0))
    || isContractWhitespace(candidate.charCodeAt(candidate.length - 1))) {
    return rejected(projectCreationPathValidationReasons.projectNameHasLeadingOrTrailingWhitespace);
  }

  if (candidate.endsWith('.')) {
    return rejected(projectCreationPathValidationReasons.projectNameEndsWithDot);
  }

  const firstDot = candidate.indexOf('.');
  const deviceNameCandidate = firstDot >= 0 ? candidate.slice(0, firstDot) : candidate;
  if (reservedDeviceNames.has(toAsciiUpperCase(deviceNameCandidate))) {
    return rejected(projectCreationPathValidationReasons.projectNameUsesReservedDeviceName);
  }

  return valid();
}

export function validateExcelWorkbookPath(candidate: string): ProjectCreationPathValidationResult {
  if (candidate.includes('[') || candidate.includes(']')) {
    return rejected(projectCreationPathValidationReasons.excelPathContainsUnsupportedCharacter);
  }

  if (candidate.length > maximumExcelWorkbookPathUtf16CodeUnitLength) {
    return rejected(projectCreationPathValidationReasons.excelPathTooLong);
  }

  return valid();
}

function valid(): ProjectCreationPathValidationResult {
  return { isValid: true, reason: null };
}

function rejected(reason: ProjectCreationPathValidationReason): ProjectCreationPathValidationResult {
  return { isValid: false, reason };
}

function isWellFormedUtf16(candidate: string): boolean {
  for (let index = 0; index < candidate.length; index++) {
    const codeUnit = candidate.charCodeAt(index);
    if (codeUnit >= 0xd800 && codeUnit <= 0xdbff) {
      if (index + 1 >= candidate.length) {
        return false;
      }

      const nextCodeUnit = candidate.charCodeAt(index + 1);
      if (nextCodeUnit < 0xdc00 || nextCodeUnit > 0xdfff) {
        return false;
      }

      index++;
    } else if (codeUnit >= 0xdc00 && codeUnit <= 0xdfff) {
      return false;
    }
  }

  return true;
}

function containsUnicodeControlCodeUnit(candidate: string): boolean {
  for (let index = 0; index < candidate.length; index++) {
    const codeUnit = candidate.charCodeAt(index);
    if (codeUnit <= 0x001f || (codeUnit >= 0x007f && codeUnit <= 0x009f)) {
      return true;
    }
  }

  return false;
}

function isContractWhitespace(codeUnit: number): boolean {
  return (codeUnit >= 0x0009 && codeUnit <= 0x000d)
    || codeUnit === 0x0020
    || codeUnit === 0x0085
    || codeUnit === 0x00a0
    || codeUnit === 0x1680
    || (codeUnit >= 0x2000 && codeUnit <= 0x200a)
    || (codeUnit >= 0x2028 && codeUnit <= 0x2029)
    || codeUnit === 0x202f
    || codeUnit === 0x205f
    || codeUnit === 0x3000;
}

function toAsciiUpperCase(value: string): string {
  let result = '';
  for (let index = 0; index < value.length; index++) {
    const codeUnit = value.charCodeAt(index);
    result += String.fromCharCode(
      codeUnit >= 0x0061 && codeUnit <= 0x007a ? codeUnit - 0x20 : codeUnit);
  }

  return result;
}

File: tools/vba-dev/src/VbaDev.Domain/ProjectCreationPathValidation.cs
namespace VbaDev.Domain;

/// <summary>
/// Stable reason identifiers for project-creation path validation version 1.0.
/// </summary>
public static class ProjectCreationPathValidationReasons
{
    public const string ProjectNameEmpty = "projectNameEmpty";
    public const string ProjectNameIllFormedUnicode = "projectNameIllFormedUnicode";
    public const string ProjectNameDotSegment = "projectNameDotSegment";
    public const string ProjectNameContainsPathSeparator = "projectNameContainsPathSeparator";
    public const string ProjectNameContainsWindowsInvalidCharacter = "projectNameContainsWindowsInvalidCharacter";
    public const string ProjectNameContainsUnicodeControlCharacter = "projectNameContainsUnicodeControlCharacter";
    public const string ProjectNameHasLeadingOrTrailingWhitespace = "projectNameHasLeadingOrTrailingWhitespace";
    public const string ProjectNameEndsWithDot = "projectNameEndsWithDot";
    public const string ProjectNameUsesReservedDeviceName = "projectNameUsesReservedDeviceName";
    public const string ExcelPathContainsUnsupportedCharacter = "excelPathContainsUnsupportedCharacter";
    public const string ExcelPathTooLong = "excelPathTooLong";
}

/// <summary>
/// Reports the first stable rejection reason from a project-creation path contract.
/// </summary>
/// <param name="Reason">The stable rejection reason, or null when the value is valid.</param>
public sealed record ProjectCreationPathValidationResult(string? Reason)
{
    /// <summary>Gets whether the candidate satisfies the contract.</summary>
    public bool IsValid => Reason is null;
}

/// <summary>
/// Validates a project name as an exact host-neutral Windows basename.
/// </summary>
public static class ProjectNameLexicalContract
{
    private static readonly HashSet<string> ReservedDeviceNames = new(
        [
            "CON",
            "PRN",
            "AUX",
            "NUL",
            "COM1",
            "COM2",
            "COM3",
            "COM4",
            "COM5",
            "COM6",
            "COM7",
            "COM8",
            "COM9",
            "COM¹",
            "COM²",
            "COM³",
            "LPT1",
            "LPT2",
            "LPT3",
            "LPT4",
            "LPT5",
            "LPT6",
            "LPT7",
            "LPT8",
            "LPT9",
            "LPT¹",
            "LPT²",
            "LPT³"
        ],
        StringComparer.OrdinalIgnoreCase);

    /// <summary>
    /// Returns the first stable rejection reason without rewriting the candidate.
    /// </summary>
    /// <param name="candidate">The exact project-name UTF-16 sequence.</param>
    /// <returns>The validation result.</returns>
    public static ProjectCreationPathValidationResult Validate(string candidate)
    {
        ArgumentNullException.ThrowIfNull(candidate);

        if (candidate.Length == 0)
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameEmpty);
        }

        if (!IsWellFormedUtf16(candidate))
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameIllFormedUnicode);
        }

        if (candidate is "." or "..")
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameDotSegment);
        }

        if (candidate.IndexOfAny(['/', '\\']) >= 0)
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameContainsPathSeparator);
        }

        if (candidate.IndexOfAny(['<', '>', ':', '"', '|', '?', '*']) >= 0)
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameContainsWindowsInvalidCharacter);
        }

        if (candidate.Any(IsUnicodeControlCodeUnit))
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameContainsUnicodeControlCharacter);
        }

        if (IsContractWhitespace(candidate[0]) || IsContractWhitespace(candidate[^1]))
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameHasLeadingOrTrailingWhitespace);
        }

        if (candidate.EndsWith(".", StringComparison.Ordinal))
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameEndsWithDot);
        }

        var firstDot = candidate.IndexOf('.', StringComparison.Ordinal);
        var deviceNameCandidate = firstDot >= 0 ? candidate[..firstDot] : candidate;
        if (ReservedDeviceNames.Contains(deviceNameCandidate))
        {
            return Rejected(ProjectCreationPathValidationReasons.ProjectNameUsesReservedDeviceName);
        }

        return new ProjectCreationPathValidationResult(null);
    }

    private static ProjectCreationPathValidationResult Rejected(string reason)
        => new(reason);

    private static bool IsWellFormedUtf16(string candidate)
    {
        for (var index = 0; index < candidate.Length; index++)
        {
            var codeUnit = candidate[index];
            if (char.IsHighSurrogate(codeUnit))
            {
                if (index + 1 >= candidate.Length || !char.IsLowSurrogate(candidate[index + 1]))
                {
                    return false;
                }

                index++;
            }
            else if (char.IsLowSurrogate(codeUnit))
            {
                return false;
            }
        }

        return true;
    }

    private static bool IsUnicodeControlCodeUnit(char codeUnit)
        => codeUnit <= '\u001f' || codeUnit is >= '\u007f' and <= '\u009f';

    private static bool IsContractWhitespace(char codeUnit)
        => codeUnit is >= '\u0009' and <= '\u000d'
            or '\u0020'
            or '\u0085'
            or '\u00a0'
            or '\u1680'
            or >= '\u2000' and <= '\u200a'
            or >= '\u2028' and <= '\u2029'
            or '\u202f'
            or '\u205f'
            or '\u3000';
}

/// <summary>
/// Validates a complete Excel-facing workbook path using version 1.0 limits.
/// </summary>
public static class ExcelWorkbookPathContract
{
    /// <summary>The inclusive Excel workbook-path limit in UTF-16 code units.</summary>
    public const int MaximumUtf16CodeUnitLength = 218;

    /// <summary>
    /// Returns the first stable rejection reason for a complete workbook path.
    /// </summary>
    /// <param name="candidate">The exact absolute Excel-facing path.</param>
    /// <returns>The validation result.</returns>
    public static ProjectCreationPathValidationResult Validate(string candidate)
    {
        ArgumentNullException.ThrowIfNull(candidate);

        if (candidate.IndexOfAny(['[', ']']) >= 0)
        {
            return new ProjectCreationPathValidationResult(
                ProjectCreationPathValidationReasons.ExcelPathContainsUnsupportedCharacter);
        }

        return new ProjectCreationPathValidationResult(
            candidate.Length > MaximumUtf16CodeUnitLength
                ? ProjectCreationPathValidationReasons.ExcelPathTooLong
                : null);
    }
}
